src/osd/OSDMap.cc

```diff
@@ -2758,7 +2758,8 @@ void OSDMap::_apply_primary_affinity(ps_t seed,
 			seed, o) >> 16) >= a) {
       // we chose not to use this primary.  note it anyway as a
       // fallback in case we don't pick anyone else, but keep looking.
-      if (pos < 0)
+      // don't use it if affinity is 0, though.
+      if (pos < 0 && a > 0)
```

Comment:
Seems right for me but would prefer to have a review also from @athanatos or @gregsfortytwo.

Reply:
I don't think we can change this without a feature flag -- it changes the osd mapping for some OSDMaps and during an upgrade OSDs would disagree about the primary.

Reply:
It's more complicated than that — this changes how the OSD will check *historical* mappings. I'm not sure if we've actually made this kind of change to how crush mappings work before — we just went to straw2 instead of trying to update straw, for instance.

Reply:
Yeah, @gregsfortytwo is right.  It's *possible* that that won't be a problem here because I can't think of a case where we care about which osd was primary for previous mappings.  Nevertheless, need to think more carefully about this.

Reply:
Or hmm, if this is only used for acting set changes which override the up set, does that mean it's persisted in the map somewhere? I'm not totally understanding the bug report. Is the issue that we get into this edge case, select the 0-affinity/out OSD as the primary, and then because it's the primary it gets hauled into the acting set when it wasn't before?

Reply:
Oh, I didn't read past noting that we can't change this method easily.  I don't think this is actually relevant to the bug report.  _apply_primary_affinity changes the up set primary, which isn't relevant to this bug.  The problem is that when we go through peering, we may set a temporary pg mapping (see PeeringState::choose_acting).  This temporary mapping bypasses crush entirely and is written explicitly into the next OSDMap.

In OSDMap::_pg_to_up_acting_osds, we invoke _get_temp_osds() before generating the up set.  _apply_primary_affininity only affects the up set primary, and therefore only the acting set primary if _get_temp_osds doesn't return a different one.

_get_temp_osds selects the first item in the acting set not equal to CRUSH_ITEM_NONE.  The core problem here is that the temp pg mapping system doesn't allow separately specifying the primary.  For replicated pgs, PeeringState::calc_replicated_acting can choose to consider primary affinity when ordering the acting set.  We're kind of out of luck for ec pgs though, since the ordering is fixed -- we'd need to specify the primary separately.

This is going to be a relatively complex change to make.

Reply:
@NitzanMordhai Did you observe that this change fixes the behavior in the bug report?

Reply:
@athanatos Yes, i tested it, and it fixes the behavior, but i didn't consider the peering issue that you explained.

Reply:
@NitzanMordhai Oh, that's actually surprising to me.  Can you outline how you tested it?  Just the test steps from the bug report?

Reply:
@NitzanMordhai I'm actually not sure the behavior you note above is worth fixing.

First, this specific fix is problematic because it may result in selecting no position to be the primary, which would be invalid.

More generally, _apply_primary_affinity will only select a weight 0 osd if that osd is first and all other osds rejected the mapping.  This is possible with your example because all of the osds other than the weight 0 one have weight .5 -- that's a very odd configuration where every osd is configured to reject some of its primary mappings (100% for osd 1 and 50% each for 0 and 2).  For ~1/4 of pgs then, _apply_primary_affinity is going to find that all osds reject being primary (1 * 1/2 * 1/2).  It's not really all that much more correct to select 0 or 2 as primary in those cases.

Reply:
@athanatos @gregsfortytwo , thanks a lot for the review. my recreation of the issue was to force the situation, and yes, it looks more complicate to fix now.

Reply:
@idryomov 's point is accurate, the client definitely cares about the primary affinity behavior, so changing it would be very complicated.  For the reasons I outlined above, no change is going to be particularly satisfying in the event that none of the osds for a PG have a primary affinity of 1, so I'm going to close this PR.